### PR TITLE
added LDAP group pruning

### PR DESCRIPTION
**Pull request by stevekuznetsov**
This pull adds `openshift ex prune-groups` and the tests associated with that. 

Depends on https://github.com/openshift/origin/pull/5708

/cc @deads2k @liggitt

### CODE REVIEW

**Comment by stevekuznetsov** [comment]
@deads2k @liggitt this is ready, and purely additions so it should be easier to reason about

[test][extended:ldap_groups]

**Comment by stevekuznetsov** [comment]
flake [here](https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin/7551/):
```
Get https://copr.fedoraproject.org/coprs/maxamillion/origin-next/repo/epel-7/maxamillion-origin-next-epel-7.repo: dial tcp 209.132.184.54:443: no route to host
```
https://github.com/openshift/origin/issues/6040

**Comment by stevekuznetsov** [comment]
re[test] [extended:ldap_groups]

**Comment by stevekuznetsov** [comment]
re[test] [extended:ldap_groups]

**Comment by stevekuznetsov** [comment]
flake on `TestAuthorizationSubjectAccessReview` [here](https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin/7560/console):
```
F1130 23:08:28.656749     704 run_components.go:393] Unable to initialize namespaces: unable to refresh the security allocation UID blocks: User "system:openshift-master" cannot list all namespaces in the cluster
```

https://github.com/openshift/origin/issues/6065

re[test][extended:ldap_groups]

File: contrib/completions/bash/openshift [hunk 1]
```diff
@@ -13883,6 +13883,77 @@ _openshift_ex_sync-groups()
...
+    flags_with_completion+=("--client-certificate")
+    flags_completion+=("_filedir")
+    flags+=("--client-key=")
+    flags_with_completion+=("--client-key")
+    flags_completion+=("_filedir")
+    flags+=("--cluster=")
+    flags+=("--config=")
+    flags_with_completion+=("--config")
+    flags_completion+=("_filedir")
+    flags+=("--container-hints=")
```
Comment (deads2k) [inline contrib/completions/bash/openshift]: I don't recognize this flag.  What's going on here?

File: pkg/cmd/experimental/syncgroups/groupdetector/groupdetector.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+	}
+
+	if len(members) == 0 {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+// NewCompoundDetector returns an LDAPGroupDetector that subsumes some other LDAPGroupDetectors
```
Comment (deads2k) [inline pkg/cmd/experimental/syncgroups/groupdetector/groupdetector.go]: This comment needs a lot of work.  I think its trying to say:

```
1.  All detectors are checked.
2.  If any detector errors, an error is returned
3.  If (no detector errors) && (any detector reports true), true is returned.
```

File: pkg/cmd/experimental/syncgroups/cli/prunegroups.go [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+	PruneGroupsRecommendedName = "prune-groups"
+
+	pruneGroupsLong = `
+Prune OpenShift Groups referencing missing records on from an external provider.
+
+In order to prune OpenShift Group records using those from an external provider, determine which Groups you wish
+to prune. For instance, all or some groups may be selected from the current Groups stored in OpenShift that have
+been synced previously. Any combination of a literal whitelist, a whitelist file and a blacklist file is supported.
+The path to a sync configuration file that was used for syncing the groups in question is required in order to 
+describe how data is requested from the external record store. Default behavior is to prune all OpenShift groups
```
Comment (deads2k) [inline pkg/cmd/experimental/syncgroups/cli/prunegroups.go]: default is to display, `--confirm` required for mutation, right?

File: pkg/cmd/experimental/syncgroups/cli/prunegroups.go [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+Prune OpenShift Groups referencing missing records on from an external provider.
+
+In order to prune OpenShift Group records using those from an external provider, determine which Groups you wish
+to prune. For instance, all or some groups may be selected from the current Groups stored in OpenShift that have
+been synced previously. Any combination of a literal whitelist, a whitelist file and a blacklist file is supported.
+The path to a sync configuration file that was used for syncing the groups in question is required in order to 
+describe how data is requested from the external record store. Default behavior is to prune all OpenShift groups 
+for which the external record does not exist.
+`
+	pruneGroupsExamples = `  # Prune all orphaned groups
```
Comment (deads2k) [inline pkg/cmd/experimental/syncgroups/cli/prunegroups.go]: don't all these simply display.  I would expect `--confirm` is needed to mutate.

File: pkg/cmd/experimental/syncgroups/cli/prunegroups.go [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+		return err
+	}
+
+	// Now we run the pruner and report any errors
+	pruneErrors := pruner.Prune()
+	return kerrs.NewAggregate(pruneErrors)
+
+}
+
+func buildPruneBuilder(clientConfig ldapclient.Config, rfc2307 *api.RFC2307Config, ad *api.ActiveDirectoryConfig, augmentedAD *api.AugmentedActiveDirectoryConfig) (PruneBuilder, error) {
```
Comment (deads2k) [inline pkg/cmd/experimental/syncgroups/cli/prunegroups.go]: I think I'm ok with passing the full `o.Config`.  If it looks reasonable, update the `buildSyncBuilder` method too.

File: pkg/cmd/experimental/syncgroups/rfc2307/ldapinterface.go [hunk 6]
```diff
@@ -62,6 +63,13 @@ type LDAPInterface struct {
 var _ interfaces.LDAPMemberExtractor = &LDAPInterface{}
 var _ interfaces.LDAPGroupGetter = &LDAPInterface{}
 var _ interfaces.LDAPGroupLister = &LDAPInterface{}
+var _ interfaces.LDAPMemberExtractor = &LDAPInterface{}
```
Comment (deads2k) [inline pkg/cmd/experimental/syncgroups/rfc2307/ldapinterface.go]: Why are there duplicates here?

File: pkg/cmd/experimental/syncgroups/rfc2307/ldapinterface.go [hunk 7]
```diff
@@ -62,6 +63,13 @@ type LDAPInterface struct {
...
+var _ interfaces.LDAPGroupGetter = &LDAPInterface{}
+var _ interfaces.LDAPGroupLister = &LDAPInterface{}
+
+func (e *LDAPInterface) String() string {
+	return fmt.Sprintf("%#v", e)
```
Comment (deads2k) [inline pkg/cmd/experimental/syncgroups/rfc2307/ldapinterface.go]: This is acyclic?

File: pkg/cmd/experimental/syncgroups/rfc2307/ldapinterface.go [hunk 8]
```diff
@@ -62,6 +63,13 @@ type LDAPInterface struct {
 var _ interfaces.LDAPMemberExtractor = &LDAPInterface{}
 var _ interfaces.LDAPGroupGetter = &LDAPInterface{}
 var _ interfaces.LDAPGroupLister = &LDAPInterface{}
+var _ interfaces.LDAPMemberExtractor = &LDAPInterface{}
+var _ interfaces.LDAPGroupGetter = &LDAPInterface{}
+var _ interfaces.LDAPGroupLister = &LDAPInterface{}
+
+func (e *LDAPInterface) String() string {
```
Comment (deads2k) [inline pkg/cmd/experimental/syncgroups/rfc2307/ldapinterface.go]: On further reflection, this probably isn't worth it.  If you want %#v, specify it.

File: contrib/completions/bash/openshift (hunk 1 above)
Comment (stevekuznetsov) [inline contrib/completions/bash/openshift]: I don't know what this is and there doesn't seem to be anything wrong / different about the `cobra

Comment (deads2k) [inline contrib/completions/bash/openshift]: > I don't know what this is and there doesn't seem to be anything wrong / different about the cobra config, so I have no idea what this flag is for.

Open an issue, it seems to be a common problem.

File: pkg/cmd/experimental/syncgroups/cli/prunegroups.go (hunk 5 above)
Comment (stevekuznetsov) [inline pkg/cmd/experimental/syncgroups/cli/prunegroups.go]: I'll add that change to the next pull in this sequence, it already contains misc. refactors

Comment (deads2k) [inline pkg/cmd/experimental/syncgroups/cli/prunegroups.go]: > I'll add that change to the next pull in this sequence, it already contains misc. refactors

What's left after this.  I was hoping to be done.

File: contrib/completions/bash/openshift (hunk 1 above)
Comment (stevekuznetsov) [inline contrib/completions/bash/openshift]: https://github.com/openshift/origin/issues/6192

**Comment by stevekuznetsov** [comment]
@deads2k addressed comments

**Comment by stevekuznetsov** [comment]
@deads2k changed `buildSyncBuilder`

**Comment by deads2k** [comment]
just missing these https://github.com/stevekuznetsov/origin/commit/6cc5b18123b9f62c4037d9a476529817abfd68f4 change, right?  Bring them in and I think this is ready.

**Comment by stevekuznetsov** [comment]
@deads2k any reason why they can't sit in their own PR like the other commits in this PR chain?

**Comment by deads2k** [comment]
> @deads2k any reason why they can't sit in their own PR like the other commits in this PR chain?

There's no reason to and they are ready now.  Prettier code is ready and available, why take anything less.

**Comment by stevekuznetsov** [comment]
@deads2k cherry-picked that into this pull

**Comment by stevekuznetsov** [comment]
[test][extended:ldap_groups]

**Comment by deads2k** [comment]
lgtm, squash for merge

**Comment by stevekuznetsov** [comment]
@deads2k squashed

**Comment by deads2k** [comment]
[merge]

**Pull request merged**